### PR TITLE
Traders send their parameters on request (+ test suites improvements)

### PR DESCRIPTION
**Pull request by merlinND**

### CODE REVIEW

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala [hunk 1]
```diff
@@ -17,6 +26,40 @@ final class ComponentBuilder(val system: ActorSystem) {
...
+        if(watched.isEmpty) cb()
+      }
+    }
+  }
+  /**
+   * @param references List of components that need to be watched
+   * @param onAllDead Callback function to call when all watched actors have been terminated
+   */
+  private case class StartKilling(references: List[ComponentRef], onAllDead: () => Any)
+  private val reaper = system.actorOf(Props(classOf[Reaper], this), "Reaper")
```
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:61]: It seems `this` is unnecessary, as reaper doesn't take parameter.

File: ts/src/main/scala/ch/epfl/ts/traders/Trader.scala [hunk 2]
```diff
@@ -41,6 +42,22 @@ abstract class Trader(val uid: Long, val parameters: StrategyParameters) extends
...
   /**
+   * @note We *do not* catch everything, we leave this partial function undefined
+   *       for all messages we cannot handle explicitly here so that concrete Traders
+   *       may handle them their own way using their `receiver` function.
+   */
+  // TODO: move all common Trader behaviors to this receiver
+  final def traderReceive: PartialFunction[Any, Unit] = {
+    case GetTraderParameters => {
+      println("Got a GetTraderParameters")
+      sender() ! parameters
```
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/traders/Trader.scala]: `sender ! parameters` would be better.

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala [hunk 3]
```diff
@@ -51,6 +94,24 @@ final class ComponentBuilder(val system: ActorSystem) {
...
     instances.head
   }
+  
+  /**
+   * Gracefully stop all managed components.
+   * When all stops are successful, we clear the `instances` list.
+   * 
+   * @return A future which completes when *all* managed actors have shut down.
+   */
+  def shutdownManagedActors(timeout: FiniteDuration = 3 seconds): Future[Unit] = {
```
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:111]: what's the relationship between `stop` and `shutdownManagedActors`? We need both, or we can just keep one?

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala (hunk 1 above)
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:61]: I was wrong, `this` is required for inner class actor.

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala [hunk 4]
```diff
@@ -17,6 +26,40 @@ final class ComponentBuilder(val system: ActorSystem) {
...
+  private class Reaper extends Actor {
+    var watched = ArrayBuffer.empty[ActorRef]
+    var cb: () => Any = () => Unit
+    
+    override def receive = {
+      case StartKilling(l, onAllDead) => {
+        cb = onAllDead
+        instances.foreach(c => {
+          context.watch(c.ar)
+          c.ar ! PoisonPill
```
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:42]: It seems there's no place where the components actually take this poison pill?

File: ts/src/main/scala/ch/epfl/ts/traders/Trader.scala [hunk 5]
```diff
@@ -41,6 +42,22 @@ abstract class Trader(val uid: Long, val parameters: StrategyParameters) extends
...
+    }
+  }
+  
+  override def receive = (componentReceive orElse traderReceive) orElse receiver
```
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/traders/Trader.scala]: what about `traderReceive orElse super.receive`?

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala (hunk 4 above)
Comment (liufengyun) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:42]: OK, this is a not a problem, actors are born to able to take poison pills.

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala (hunk 1 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:61]: Yes, I took a while to figure it out!

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala (hunk 4 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:42]: :pill:

File: ts/src/main/scala/ch/epfl/ts/traders/Trader.scala (hunk 5 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/traders/Trader.scala]: That's a neat idea, thanks!

File: ts/src/main/scala/ch/epfl/ts/component/Component.scala (hunk 3 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/Component.scala:111]: This is a delicate point, and I think that the difference is:
- `start` / `stop` could be called several times (like restarting actors without actually killing them)
- `shutdown` really kills the actor

I will try to make it clear in the documentation.

**Pull request merged**